File: inst/d3/scatterplot/horzbarchart.js
let margin = {
  top: 50,
  right: 10,
  bottom: 50,
  left: 100,
};

data.sort(function (a, b) {
  if (options.sort === "none") {
    return null;
  } else if (options.sort === "ascending") {
    return d3.ascending(a[options.value], b[options.value]);
  } else {
    return d3.descending(a[options.value], b[options.value]);
  }
});

let x = d3
  .scaleLinear()
  .domain([0, d3.max(data, (d) => d[options.value])])
  .range([margin.left, width - margin.right]);

let y = d3
  .scaleBand()
  .domain(d3.range(data.length))
  .rangeRound([margin.top, height - margin.bottom])
  .padding(options.paddingWidth);

let xAxis = (g) =>
  g
    .attr("transform", `translate(0, ${height - margin.bottom})`)
    .call(d3.axisBottom(x).ticks(options.valueTicks))
    .call((g) => g.select(".domain").remove());

let yAxis = (g) =>
  g
    .attr("transform", `translate(${margin.left}, 0)`)
    .call(d3.axisLeft(y).tickFormat((i) => data[i][options.label]));

svg
  .attr("viewBox", [0, 0, width, height])
  .style("background-color", options.bgcol);

svg
  .append("g")
  .attr("fill", options.fill)
  .selectAll("rect")
  .data(data)
  .enter()
  .append("rect")
  .attr("x", x(0))
  .attr("y", (d, i) => y(i))
  .attr("width", (d) => x(d[options.value]) - x(0))
  .attr("height", y.bandwidth())
  .attr("opacity", options.opacity)
  .attr("stroke", options.stroke)
  .attr("stroke-width", options.strokeWidth);

svg.append("g").call(xAxis).attr("font-size", options.valueFontSize);

svg
  .append("g")
  .call(yAxis)
  .attr("font-size", options.labelFontSize)
  .style("font-family", options.font);

// rendering the x-axis title

svg
  .append("text")
  .attr("transform", "translate(" + width / 2 + " ," + (height - 3) + ")")
  .attr("dx", "0em")
  .attr("dy", "-0.4em")
  .style("text-anchor", "middle")
  .style("font-family", options.font)
  .style("font-size", options.valueTitleFontSize)
  .text(options.valueTitle)
  .style("fill", options.axisCol);

// Rendering the y-axis title
svg
  .append("text")
  .attr("transform", "translate(" + 2 + " ," + height / 2 + ") rotate(-90)")
  .attr("dy", "1em")
  .style("text-anchor", "middle")
  .style("font-size", options.labelTitleFontSize)
  .style("font-family", options.font)
  .text(options.labelTitle)
  .style("fill", options.axisCol);

// Rendering the chart title
svg
  .append("text")
  .attr("x", width / 2)
  .attr("y", margin.top / 2)
  .attr("text-anchor", "middle")
  .attr("dx", "0em")
  .style("font-size", options.titleFontSize)
  .style("font-family", options.font)
  .text(options.title)
  .style("fill", options.axisCol);

// Modifying the color of the ticks and labels
svg.selectAll(".tick line").attr("stroke", options.axisCol);
svg.selectAll(".tick text").style("fill", options.axisCol);
svg.selectAll("path.domain").attr("stroke", options.axisCol);

File: R/horzBarChart.R

#' Create a horizontal bar chart
#'
#' @param data The data frame containing the variables to consider.
#' @param value The numeric variable to consider. Will be plotted on the x-axis.
#' @param label The categorical variable to consider. Will be plotted on the y-axis.
#' @param fill The color of the bars. Defaults to 'crimson'.
#' @param stroke Optional. The color of the stroke of the bars.
#' @param strokeWidth The width of the stroke of the bars. Defaults to 1 when the 'stroke' parameter is used.
#' @param paddingWidth The distance between each bar.
#' The value goes from 0 to 0.99 included. Defaults to 0.1.
#' @param valueTicks Optional. the number of x-axis ticks to consider.
#' @param valueFontSize  The font size of the x-axis values. Defaults to 10.
#' @param labelFontSize The font size of the y-axis labels. Defaults to 10.
#' @param valueTitle Optional. The title of the x-axis.
#' @param valueTitleFontSize The font size of the x-axis title if specified. Defaults to 14.
#' @param labelTitle Optional. The title of the y-axis.
#' @param labelTitleFontSize The font size of the y-axis title. Defaults to 14.
#' @param font The font family of the text. Defaults to "Verdana, Geneva, Tahoma, sans-serif"
#' @param title Optional. The title of the overall graphic.
#' @param titleFontSize The font size of the overall graphic's title when specified.
#' @param opacity The color opacity of the bars. Goes from 0 to 1. Defaults to 1.
#' @param axisCol the color of the x and y axis. It includes the ticks, the labels and titles.
#' Defaults to 'black'.
#' @param width Optional. The width of the SVG output.
#' @param height Optional. The height of the SVG output.
#' @param sort Optional. Takes the following arguments:
#' 'none', 'ascending' or 'descending', default to 'none'
#'
#' @param bgcol Optional. The color of the background, default to:
#' '#CAD0D3'
#'
#' @return A SVG horizontal bar chart.
#' @export
#'
#' @examples
#' library(ggplot2) # needed for the mpg data frame
#' library(dplyr) # needed for the data wrangling process
#'
#' mpg %>% group_by(manufacturer) %>%
#'  summarise(median_hwy = median(hwy)) %>%
#'  horzBarChart(
#'    label = "manufacturer",
#'    value = "median_hwy",
#'    sort = "ascending"
#'  )

horzBarChart <- function(data,
                         label,
                         value,
                         fill = "crimson",
                         sort = "none",
                         paddingWidth = 0.1,
                         stroke = NULL,
                         strokeWidth = 1,
                         bgcol = "#CAD0D3",
                         valueTicks = NULL,
                         valueFontSize = 10,
                         labelFontSize = 10,
                         valueTitle = NULL,
                         valueTitleFontSize = 14,
                         labelTitle = NULL,
                         labelTitleFontSize = 14,
                         font = "Verdana, Geneva, Tahoma, sans-serif",
                         title = NULL,
                         titleFontSize = 20,
                         opacity = 1,
                         axisCol = "black",
                         width = NULL,
                         height = NULL
                         ) {


  if (is.null(data[[value]]) || is.null(data[[label]])) {
    stop("Please check that x and y belong to the specified data frame")
  }

  if (!(any(c('none',
              'ascending',
              'descending') %in% sort))) {
    stop("
    the sort parameter can only take the following values:
    'none', 'ascending' or 'descending'
       ")
  }

  if (grepl(";", font)) {
    stop("please remove the ';' character from your font argument")
  }


  r2d3::r2d3(
    data = data,
    script = system.file(
      "d3/scatterplot/horzbarchart.js",
      package = "ddplot"
    ),
    options = list(
      value = value,
      label = label,
      fill = fill,
      paddingWidth = paddingWidth,
      valueTicks = valueTicks,
      valueFontSize = valueFontSize,
      labelFontSize = labelFontSize,
      valueTitle = valueTitle,
      valueTitleFontSize = valueTitleFontSize,
      labelTitle = labelTitle,
      labelTitleFontSize = labelTitleFontSize,
      font = font,
      title = title,
      titleFontSize = titleFontSize,
      stroke = stroke,
      strokeWidth = strokeWidth,
      opacity = opacity,
      width = width,
      height = height,
      sort = sort,
      bgcol = bgcol,
      axisCol = axisCol
    )
  )


}
